LoftQ: Add LoftQ method integrated into LoRA. Add example code for LoftQ usage.

yxli2123 opened the pull request:
I have added the LoftQ method to LoRA. I have run the `make style` command.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_1150). All of your documentation changes will be reflected on that endpoint.

BenjaminBossan commented on examples/loftq_finetuning/train_gsm8k_llama.py:
```diff
@@ -0,0 +1,871 @@
+
+def extract_answer_number(sentence: str) -> float:
+    sentence = sentence.replace(",", "")
+    pred = [s for s in re.findall(r"-?\d+\.?\d*", sentence)]
```
Apparently, ruff complains that `[s for s in re.findall(r"-?\d+\.?\d*", sentence)]` should be `list(re.findall(r"-?\d+\.?\d*", sentence))`. Fine I guess, but when I tried, `findall` already returns a list, so neither should be necessary.

What you could do on top is pre-compile the regex. So something like

```python
PATTERN_NUMBER = re.compile(r"-?\d+\.?\d*")

def extract_answer_number(sentence: str) -> float:
    sentence = sentence.replace(",", "")
    pred = PATTERN_NUMBER.findall(sentence)
    ...
        pred_answer = PATTERN_NUMBER.findall(segment[1])
        ...
```

This should be a bit faster and avoid repeating the same regex.

yxli2123 commented:
Is it possible to add `scipy` package to the dependency? I may need this package to implement norm.ppf function.

BenjaminBossan commented:
> I failed the test because `scipy` is not installed. Is it possible to add `scipy` package to the dependency? I may need this package to implement norm.ppf function.

Ah, too bad that this isn't implemented in PyTorch. I would suggest not to add scipy as a standard dependency, as we want to keep those lean. But we can add it as a [dev dependency](https://github.com/huggingface/peft/blob/043238578f9c7af88334005819543cf60e263760/setup.py#L20) and use a local import of `scipy.stats.norm` inside of `create_normal_map`.
On top of that, let's alert users who want to use LoftQ as early as possible if they don't have scipy installed. For this, we can check in the `__post_init__` of `LoraConfig` that scipy is installed if `self.init_lora_weights == "loftq"`.

Alternatively, we could add our own `norm.ppf` function to PEFT. I checked the scipy code and it's unfortunately not straightforward to just copy it. It seems to be based on `special.ndtri`, which PyTorch also implements, but I'd have to brush up my math to check how to do this correctly. At least on surface, it seems possible:

![image](https://github.com/huggingface/peft/assets/6229650/22bbf71a-158b-4d41-bea3-4a3026ee8346)

(`ndtri` leads to many NaNs though)

WDYT?

pacman100 commented on examples/loftq_finetuning/README.md:
```diff
@@ -0,0 +1,68 @@
+
+## Introduction
+
+LoftQ provides better initialization for LoRA adaptors A and B, 
```
```suggestion
LoftQ provides better initialization for LoRA adapters A and B, 
```

pacman100 commented on src/peft/tuners/lora/config.py:
```diff
@@ -20,6 +20,42 @@
+    )
+    loftq_bits: int = field(default=4, metadata={"help": "Quantization bits for LoftQ"})
+    loftq_iter: int = field(default=1, metadata={"help": "Alternating iterations for LoftQ"})
+    loftq_fake: bool = field(
```
This isn't used anywhere in `loftq_utils.py`, is it?

yxli2123 replied:
The `loftq_bits` and `loftq_iter` are used. `loftq_fake` and `bits_pattern` are not used, but could be the future feature. Anyway, I have deleted the `loftq_fake` and `bits_pattern`.

pacman100 reviewed (commented):
Thank you @yxli2123 for super cool work wrt adding LoftQ initialization method for LoRA adapters. 

wrt `scipy` error, we can overcome by importing `from peft.utils.loftq_utils import loftq_init` in the `def loftq_init(self, adapter_name)` method wherein we first raise and error if scipy is not installed. 

Apart from that, left a nit suggestion and a question regarding `loftq_fake`.

yxli2123 commented:
Hi, thanks for providing constructive suggestions. I have edited the code. Could you review the code and run test?

BenjaminBossan reviewed (changes requested):
Thanks a lot, I think we're almost finished with the PR.

Could you please run `make style` so that CI passes? Also, I found a few minor issues, please check my comments. Finally, the license snippet is missing on some files, would you please be so kind to add it? Thanks!

yxli2123 commented:
Thanks again. I have added license to `loftq_utils.py`, changed `assert` style to `if ... raise ValueError`, removed unnecessary `print`, moved the `scipy` check to `__post_init__`, and run `make style`.

BenjaminBossan reviewed (approved):
Thanks a lot for addressing the remaining comments. This is a great PR and a very useful feature to have in PEFT.

For me, the PR is in a state that it can be merged. I'll check if we want to have another final review by a colleague. For the future, we also should add documentation and unit tests. If no one else wants to work on those, I'll try to find some time next week to tackle that.

pacman100 reviewed (approved):
Great job @yxli2123 on adding LoftQ and addressing all the comments! 🔥🚀✨

BenjaminBossan commented:
Ah, sorry @yxli2123 I merged #1189, which created some merge conflicts, I should have merged this PR first. Would you be so kind to fix them? Otherwise, LMK if I should push the fixes on top of your PR.

yxli2123 commented:
I have edited my branch with the conflicts. It should be fine to just accept my branch when conflicts happen.

BenjaminBossan commented on src/peft/tuners/lora/config.py:
```diff
@@ -86,6 +105,7 @@ class LoraConfig(PeftConfig):
                 "initialization from the reference implementation from Microsoft. Passing 'gaussian' results "
                 "in Gaussian initialization scaled by the LoRA rank for linear and layers. Setting the initialization "
                 "to False leads to completely random initialization and is discouraged."
+                "Pass `'loftq'` to use LoftQ initialization"
```
Above, could you please add `"loftq"` to the options: `Literal["gaussian", "loftq"]`

yxli2123 commented:
Thank you! We are glad to contribute to this wonderful open-source project.
